### PR TITLE
fix(router): check if windows.console exists before using it

### LINKED ISSUE #11487: Router uses console.group() and console.groupEnd() without checking if they exist
**I'm submitting a ...**  (check one with "x")

```
[X] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

In [angular/modules/@angular/router/src/router_module.ts](https://github.com/angular/angular/blob/master/modules/%40angular/router/src/router_module.ts), line 173-176, `console.group()` and `console.groupEnd()` are used without checking if they exist.  This checking _should be_ performed (as it is elsewhere):

```
if (console.group) { console.group("..."); }
...
if (console.groupEnd) { console.groupEnd("..."); }
```

**Current behavior** 
Router throws error on browsers that don't have `console.group` and `console.groupEnd`

**Expected behavior**
No attempt is made to group() or groupEnd() on browsers that don't support it.

**Reproduction of the problem**
Try on IE 10

**What is the motivation / use case for changing the behavior?**
Client wants to run application on IE 10.  It works fine except for this problem and some internationalization stuff.

**Please tell us about your environment:**
Windows 7, IE 10
- **Angular version:** 2.0.0-rc.6
  Still an issue in rc6
- **Browser:** IE 10
- **Language:** TypeScript 1.8
- **Node (for AoT issues):** `node --version` =

### PR DESCRIPTION
**Pull request by DzmitryShylovich**
Fixes https://github.com/angular/angular/issues/11487

### CODE REVIEW

**Comment by vicb** [comment]
please rebase. Thanks

**Comment by vicb** [comment]
Oops sorry, out of sync again, could you rebase one more time ?

File: modules/@angular/router/src/router_module.ts [hunk 1]
```diff
@@ -251,10 +251,16 @@ export function setupRouter(
 
   if (opts.enableTracing) {
     router.events.subscribe(e => {
-      console.group(`Router Event: ${(<any>e.constructor).name}`);
-      console.log(e.toString());
-      console.log(e);
-      console.groupEnd();
+      if (window.console) {
```
Comment (laco0416) [inline modules/@angular/router/src/router_module.ts]: this code can be broken in the case below

```js
// node.js context

global.window = {
  console: {}
}
```

I think it should check the `global` variable and use `global.console` if it exists.

Comment (laco0416) [inline modules/@angular/router/src/router_module.ts]: or something like consoleService...

Comment (DzmitryShylovich) [inline modules/@angular/router/src/router_module.ts]: http://prntscr.com/cwzv5l ?

Comment (laco0416) [inline modules/@angular/router/src/router_module.ts]: node context.

Comment (DzmitryShylovich) [inline modules/@angular/router/src/router_module.ts]: Probably I should use https://github.com/angular/angular/blob/41c8c309730d36e46012f6461e8c94d84b13554d/modules/%40angular/core/src/console.ts#L13

**Comment by DzmitryShylovich** [comment]
Related to https://github.com/angular/angular/issues/12169

**Comment by vsavkin** [comment]
You should use `DomAdapter.logGroup`. It'll handle all the cases when the group method is present. So you won't have to write your code using the defensive style.

**Comment by DzmitryShylovich** [comment]
@vsavkin ok, how I can import `getDOM` method in router module?
```
import {getDOM} from '@angular/platform-browser/src/dom/dom_adapter';

or

import {__platform_browser_private__} from '@angular/platform-browser';
__platform_browser_private__.getDOM();
```
neither of these works

File: modules/@angular/platform-browser/src/browser/browser_adapter.ts [hunk 2]
```diff
@@ -77,16 +77,16 @@ export class BrowserDomAdapter extends GenericBrowserDomAdapter {
   invoke(el: Node, methodName: string, args: any[]): any { (<any>el)[methodName](...args); }
 
   // TODO(tbosch): move this into a separate environment class once we have it
-  logError(error: string) { (window.console.error || window.console.log)(error); }
+  logError(error: string) { window.console && (window.console.error || window.console.log)(error); }
```
Comment (vicb) [inline modules/@angular/platform-browser/src/browser/browser_adapter.ts]: please make `if (windows.console) {...}` explicit for readability

**Comment by vicb** [comment]
Please integrate the comments and squash/rebase.

To import "private" stuff, check https://github.com/angular/angular/blob/master/modules/%40angular/platform-browser/src/private_import_core.ts

**Comment by DzmitryShylovich** [comment]
@vicb https://github.com/angular/angular/pull/12348/files#diff-0ba2e46ad7156d0a763f63a4c355b1e6R1 still throws `Cannot find module '@angular/platform-browser'.`

File: modules/@angular/platform-browser/src/browser/browser_adapter.ts [hunk 3]
```diff
@@ -77,16 +77,30 @@ export class BrowserDomAdapter extends GenericBrowserDomAdapter {
   invoke(el: Node, methodName: string, args: any[]): any { (<any>el)[methodName](...args); }
 
   // TODO(tbosch): move this into a separate environment class once we have it
-  logError(error: string) { (window.console.error || window.console.log)(error); }
+  logError(error: string): void {
+    if (window.console) {
```
Comment (vicb) [inline modules/@angular/platform-browser/src/browser/browser_adapter.ts]: :+1:

File: modules/@angular/router/src/private_import_platform-browser.ts [hunk 4]
```diff
@@ -0,0 +1,3 @@
+import {__platform_browser_private__ as r} from '@angular/platform-browser';
```
Comment (vicb) [inline modules/@angular/router/src/private_import_platform-browser.ts]: please add the header

**Comment by vicb** [comment]
Could you make sure all the `tsconfig.json` references platform-browser.

@vsavkin any other idea of what could go wrong ?

**Comment by DzmitryShylovich** [comment]
@vicb done

**Comment by vicb** [comment]
Great, thanks again !

File: modules/@angular/platform-browser/src/browser/browser_adapter.ts (hunk 3 above)
Comment (DzmitryShylovich) [inline modules/@angular/platform-browser/src/browser/browser_adapter.ts]: https://github.com/angular/angular/issues/12169

**Pull request merged**